### ISSUE TITLE
react/prop-types: false positive for generic `props` usage within React class 

### ISSUE THREAD

**Issue by JHilker** [yannickcr/eslint-plugin-react#2457]
When `props` is used as an argument for a generic method within a component, its validated against the base components PropTypes. I would expect it to either be ignored from the check, or considered it's own component.

The follow code works for `7.14.3`, however fails in `7.15.0` and later with

```
'children' is missing in props validation eslint(react/prop-types)
```

```js
import React, { Component, PropTypes } from 'react';

export default class Foo extends Component {
  static get propTypes() {
    return {
      bar: PropTypes.func.isRequired,
    };
  }

  createBar() {
    return props => {
      const { children } = props;
      return <div>{children}</div>;
    };
  }

  render() {
    const Bar = this.createBar();
    return (
      <Bar>
        <span bar={this.props.bar} />
      </Bar>
    );
  }
}
```

Rule config:
```
"react/prop-types": [
  "error",
  {
    "ignore": [],
    "customValidators": [],
    "skipUndeclared": true,
  },
],
```

So far the cases I've come across where this was a problem should probably be refactored to a separate component which would sidestep the issue, but it definitely seems to be a regression to the rule regardless.

**Comment by ljharb** [comment]
`createBar` is returning a component - a function that takes props and returns jsx. That should a) be hoisted outside of `createBar`, and b) since it's a component, it needs propTypes.

**Comment by JHilker** [comment]
Hey @ljharb, thanks for getting back to me!

> That should a) be hoisted outside of createBar

Yep, I totally agree with this sentiment. I'm just reporting this because the line the rule is reporting seems inaccurate within the context of the rule. Even if you had `propTypes` on the internal component like so:

```js
createBar() {
  const Bar = props => {
    const { children } = props;
    return <div>{children}</div>;
  };

  Bar.propTypes = {
    children: ...
  };

  return Bar;
}
```

The rule would report in the same way, which could lead to engineers being confused because they did declare `propTypes` on the internal component, but the rule is stating they did not. I believe before this change the behavior was that these prop-types weren't validated at all in either case, which I'd argue is a bit better than having false positives.

The new behavior is also a bit inconsistent, since if you don't explicitly use `props`, and instead destructure the functional component props, the rule does still pass like it did previously. Example:

```js
createBar() {
  return ({ children }) => {
    return <div>{children}</div>;
  };
}
```

If restoring the original behavior where the nested component props weren't validated would add too much complexity to the rule, then I'm totally fine if this is issue is closed as a wont-fix. It is definitely an edge case due to being an already undesirable coding pattern.

I suppose the "right" answer here is likely a `no-nested-components` rule which would error in addition to this one, any idea if a similar rule already exists?

Just for context, I'm currently rolling out an upgrade of the extension to ~600 repos right now across my organization, so this is just a bit of a manual hurdle on an otherwise nearly automatable change. Though with the other excellent improvements to the rule such as supporting nested destructuring which will also require some manual fixes, it's a bit hard to gauge the exact impact of this specific regression.

> b) since it's a component, it needs propTypes.

Technically we have `"skipUndeclared": true,` specified, since we're currently ok with no props being specified (other type checking, etc.)

**Comment by ljharb** [comment]
Interesting. Clearly we need many more test cases here as part of fixing this :-)

**Comment by jzabala** [comment]
Functions starting with a lowercase letter are not considered components anymore. This should be fixed.

@ljharb
